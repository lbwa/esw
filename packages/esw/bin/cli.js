#!/usr/bin/env node

// eslint-disable-next-line no-unused-expressions
require('../src').default(process.cwd()).parse(process.argv.slice(2))
