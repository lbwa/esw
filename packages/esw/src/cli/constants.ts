export enum AvailableCommands {
  Build = 'build',
  Watch = 'watch'
}
