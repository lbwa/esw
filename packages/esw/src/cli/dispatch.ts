import path from 'path'
import arg from 'arg'
import {
  EMPTY,
  from,
  of,
  throwError,
  zip,
  tap,
  map,
  catchError,
  Observable,
  concatMap,
  switchMap
} from 'rxjs'
import { PackageJson } from 'type-fest'
import { stdout, ExitCode } from '@eswjs/common'

export type CommandRunner<V = unknown> = (argv?: string[]) => Observable<V>
type Commands = typeof AVAILABLE_COMMANDS
type CommandNames = keyof Commands

// eslint-disable-next-line @typescript-eslint/no-var-requires
const pkgJson = require(path.resolve(
  __dirname,
  '../..',
  'package.json'
)) as PackageJson
const AVAILABLE_COMMANDS = {
  build: () => from(import('../build/cli')).pipe(map(({ default: run }) => run))
}
const availableArgs = {
  '--version': Boolean,
  '--help': Boolean,

  // alias
  '-v': '--version',
  '-h': '--help'
}

function printUsageIntoTerminal(commands: Commands) {
  const names = Object.keys(commands)
  stdout.raw(
    `
Usage
  $ esw <command>

Available commands
  ${names.join(', ')}

Options
  --version, -v      Show version number
  --help, -h         Display help messages

For more information run a command with the --help flag
  $ esw ${names[0] ?? ''} --help

`,
    ExitCode.OK
  )
}

/**
 * @example parse(process.argv.slice(2)).subscribe({ next: handleArgs })
 */
export default function parse(argv: string[]) {
  const argv$ = of(argv)
  const resolvedArgv$ = argv$.pipe(
    map(argv =>
      arg(availableArgs, {
        //https://github.com/vercel/arg/blob/5.0.0/index.js#L13
        argv,
        permissive: true
      })
    ),
    catchError((err: Error & { code: string }) => {
      if (err.code === 'ARG_UNKNOWN_OPTION') {
        printUsageIntoTerminal(AVAILABLE_COMMANDS)
        return EMPTY
      }
      return throwError(() => err)
    })
  )

  const handlePrintVersion$ = resolvedArgv$.pipe(
    switchMap(args => {
      if (args['--version']) {
        stdout.raw(`v${pkgJson.version}\n`, ExitCode.OK)
        return EMPTY
      }
      return of(args)
    })
  )

  const handlePrintUsage$ = handlePrintVersion$.pipe(
    concatMap(args => {
      const { _: [commandStdin] = [] } = args
      const isValidStdin = !!AVAILABLE_COMMANDS[commandStdin as CommandNames]
      if (isValidStdin) {
        return of({ isValidStdin, args })
      }

      /**
       * 1. not a valid stdin,
       * 2. with a global help flag
       */
      printUsageIntoTerminal(AVAILABLE_COMMANDS)
      return EMPTY
    })
  )

  const handleForwardArgs$ = handlePrintUsage$.pipe(
    map(({ isValidStdin, args }) => {
      const commandName = args._[0]
      const forwardArgs = isValidStdin ? args._.slice(1) : args._
      if (args['--help']) {
        forwardArgs.push('--help')
      }
      return {
        forwardArgs,
        commandName
      }
    })
  )

  const setEnvVariables$ = handleForwardArgs$.pipe(
    tap(({ commandName }) => {
      const normalizedEnv =
        commandName === 'build' ? 'production' : 'development'
      process.env['NODE_ENV'] ||= normalizedEnv

      // Make sure commands gracefully respect termination signals (e.g. from Docker)
      process.on('SIGTERM', () => process.exit(ExitCode.OK))
      process.on('SIGINT', () => process.exit(ExitCode.OK))
    })
  )

  const handleCommandStdin$ = setEnvVariables$.pipe(
    concatMap(({ commandName, forwardArgs }) =>
      zip(
        from(AVAILABLE_COMMANDS[commandName as CommandNames]()),
        of(forwardArgs)
      )
    ),
    concatMap(([run, args]) => run?.(args))
  )

  return handleCommandStdin$
}
