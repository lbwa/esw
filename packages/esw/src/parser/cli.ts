import path from 'path'
import arg from 'arg'
import {
  EMPTY,
  from,
  of,
  throwError,
  zip,
  tap,
  map,
  catchError,
  Observable,
  concatMap
} from 'rxjs'
import { PackageJson } from 'type-fest'
import { printToTerminal } from '../shared/printer'
import { ProcessCode as Code } from '../shared/constants'

export type CommandRunner<V = unknown> = (argv?: string[]) => Observable<V>
type Commands = typeof AVAILABLE_COMMANDS
type CommandNames = keyof Commands

// eslint-disable-next-line @typescript-eslint/no-var-requires
const pkgJson = require(path.resolve(
  __dirname,
  '../..',
  'package.json'
)) as PackageJson
const AVAILABLE_COMMANDS = {
  build: () => from(import('../cli/build')).pipe(map(({ default: run }) => run))
}
const availableArgs = {
  '--version': Boolean,
  '--help': Boolean,

  // alias
  '-v': '--version',
  '-h': '--help'
}

function printUsageIntoTerminal(commands: Commands) {
  const names = Object.keys(commands)
  printToTerminal(
    `
    Usage
      $ esw <command>

      Available commands
        ${names.join(', ')}

      Options
        --version, -v Show version number
        --help, -h Display help messages

      For more information run a command with the --help flag
        $ esw ${names[0] ?? ''} --help
  `,
    Code.OK
  )
}

/**
 * @example parse(process.argv.slice(2)).subscribe({ ... })
 */
export default function parse(argv: string[]) {
  const argv$ = of(argv)
  const resolvedArgv$ = argv$.pipe(
    map(argv =>
      arg(availableArgs, {
        //https://github.com/vercel/arg/blob/5.0.0/index.js#L13
        argv,
        permissive: true
      })
    ),
    catchError((err: Error & { code: string }) => {
      if (err.code === 'ARG_UNKNOWN_OPTION') {
        printUsageIntoTerminal(AVAILABLE_COMMANDS)
        return EMPTY
      }
      return throwError(() => err)
    })
  )

  const handlePrintVersion$ = resolvedArgv$.pipe(
    tap(args => {
      if (args['--version']) {
        printToTerminal(`v${pkgJson.version}`, Code.OK)
      }
    })
  )

  const handlePrintUsage$ = handlePrintVersion$.pipe(
    concatMap(args => {
      const { _: [commandStdin] = [], '--help': isHelpStdin } = args
      const isValidStdin = !!AVAILABLE_COMMANDS[commandStdin as CommandNames]
      if (!isValidStdin || isHelpStdin) {
        printUsageIntoTerminal(AVAILABLE_COMMANDS)
        return EMPTY
      }
      return of({ isValidStdin, args })
    })
  )

  const handleForwardArgs$ = handlePrintUsage$.pipe(
    map(({ isValidStdin, args }) => {
      const commandName = args._[0]
      const forwardArgs = isValidStdin ? args._.slice(1) : args._
      if (args['--help']) {
        forwardArgs.push('--help')
      }
      return {
        forwardArgs,
        commandName
      }
    })
  )

  const setEnvVariables$ = handleForwardArgs$.pipe(
    tap(({ commandName }) => {
      const normalizedEnv =
        commandName === 'build' ? 'production' : 'development'
      process.env['NODE_ENV'] ||= normalizedEnv

      process.on('SIGTERM', () => process.exit(0))
      process.on('SIGINT', () => process.exit(0))
    })
  )

  const handleCommandStdin$ = setEnvVariables$.pipe(
    concatMap(({ commandName, forwardArgs }) =>
      zip(
        from(AVAILABLE_COMMANDS[commandName as CommandNames]()),
        of(forwardArgs)
      )
    ),
    concatMap(([run, args]) => run?.(args))
  )

  return handleCommandStdin$
}
