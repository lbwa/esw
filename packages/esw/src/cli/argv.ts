import { debuglog } from 'util'
import { concatMap, iif, map, Observable, of, tap } from 'rxjs'
import { CliSpec, createCliParser } from './parser'

const debug = debuglog('esw:argv')

export function resolveArgv<Spec extends CliSpec>(
  argv: string[] = [],
  spec: Spec,
  printUsage$: Observable<never>
) {
  const argv$ = of(argv)
  const resolvedArgv$ = argv$.pipe(
    map(argv => createCliParser(argv).parse(spec))
  )

  return resolvedArgv$.pipe(
    concatMap(argv => iif(() => !!argv['--help'], printUsage$, of(argv))),
    tap(result => debug('%o', result))
  )
}
