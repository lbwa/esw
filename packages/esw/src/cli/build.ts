import { mkdirSync, writeFileSync } from 'fs'
import path from 'path'
import arg from 'arg'
import {
  tap,
  map,
  from,
  partition,
  share,
  mergeMap,
  reduce,
  defer,
  NEVER
} from 'rxjs'
import omit from 'lodash/omit'
import cloneDeep from 'lodash/cloneDeep'
import { BuildFailure, BuildOptions, BuildResult, Metafile } from 'esbuild'
import {
  isDef,
  serializeSize,
  printTable,
  printBuildError,
  stdout,
  ExitCode
} from '@eswjs/common'
import { CommandRunner } from '../parser/cli'
import { Build } from '../build'
import { BuildArgsSpec, BUILD_ARGS_SPEC } from '../shared/cli-spec'
import { serializeArgv } from '../observable/argv'

function createPrintUsage$(exitCode = ExitCode.OK) {
  return defer(() => {
    stdout.raw(
      `
Description
  Compiles the codebase for publishing npm package.

Usage
  $ esw build [entry files]

  [entry file] represents the library entry point.
  If no entry is provided, the basename from main and module field in package.json will be used.
  User should always specific a entry point explicitly when the main and module have a different basename.

`,
      exitCode,
      false
    )
    return NEVER
  })
}

function fulfillBuildResultFilter(
  result: PromiseSettledResult<BuildResult>
): result is PromiseFulfilledResult<BuildResult> {
  return result.status === 'fulfilled'
}

async function writeToDisk(
  outPath: string,
  content: string | NodeJS.ArrayBufferView
) {
  mkdirSync(path.dirname(outPath), { recursive: true })
  writeFileSync(outPath, content, { encoding: null })
}

const build: CommandRunner<ExitCode> = function (argv = []) {
  const handlePrintUsage$ = serializeArgv(
    argv,
    BUILD_ARGS_SPEC,
    createPrintUsage$()
  )

  const normalizedBuildArgs$ = handlePrintUsage$.pipe(
    map((args: arg.Result<BuildArgsSpec>) => {
      args['--absWorkingDir'] ??= process.cwd()
      const entryPoints = args._.filter(pending => !pending.startsWith('-'))
      args['--entryPoints'] ??= entryPoints.length > 0 ? entryPoints : undefined
      return omit(args, '_')
    }),
    map(args =>
      Object.keys(args).reduce((options, key) => {
        const value = args[key as keyof BuildArgsSpec]
        // we use null to mark illegal command line value
        if (isDef(value)) {
          const name = key.replace(/^-+/, '') as keyof BuildOptions
          // @ts-ignore FIXME
          options[name] = value
        }
        return options
      }, {} as BuildOptions)
    )
  )

  let build: Build
  const handleBuilding$ = normalizedBuildArgs$.pipe(
    mergeMap(options => {
      build = new Build(options)
      return build.run()
    }),
    mergeMap(allResults => from(allResults)),
    share()
  )
  const [handleBuildResult$, handleExceptionResult$] = partition(
    handleBuilding$,
    fulfillBuildResultFilter
  )

  handleExceptionResult$
    .pipe(
      tap(({ reason }) => {
        process.exitCode = ExitCode.ERROR
        printBuildError(reason as BuildFailure)
      })
    )
    .subscribe(() => process.exit())

  const builtRecord = new Set<string>()

  const handleWriteOutFiles$ = handleBuildResult$.pipe(
    reduce((metaFiles, { value: buildResult = {} }) => {
      const { outputFiles = [], metafile = {} as Metafile } = buildResult
      outputFiles.forEach(({ path: outPath /* absolute path */, contents }) => {
        const destinationPath = build?.pathsMap
          ?.get(outPath)
          ?.find(item => !builtRecord.has(item))
        if (destinationPath) builtRecord.add(destinationPath)
        const destination = destinationPath ?? /* eg. splitted chunks */ outPath

        const cloned = cloneDeep(metafile)
        if (!outPath.endsWith(destination)) {
          const matchedDest = Object.keys(cloned.outputs).find(file =>
            outPath.endsWith(file)
          )
          if (matchedDest) {
            /**
             * @description re-defined matched dest key in metafile.outputs
             * when we use internal path, instead of esbuild output path
             */
            cloned.outputs[destination] = cloned.outputs[
              matchedDest
            ] as NonNullable<Metafile['outputs'][string]>
            delete cloned.outputs[matchedDest]
          }
        }

        void writeToDisk(destination, contents)
        metaFiles.push(cloned)
      }, [] as string[])
      return metaFiles
    }, [] as Metafile[]),
    map(metaFiles => {
      builtRecord.clear()
      if (metaFiles.length < 1) {
        stdout.warn("The file wasn't created.")
        return ExitCode.ERROR
      }

      const outputs = metaFiles.reduce(
        (group, { outputs }) => Object.assign(group, outputs),
        {} as Metafile['outputs']
      )

      const message = [
        ['Files', 'Size'],
        ...Object.keys(outputs).map(filename => [
          filename,
          isDef(outputs[filename]?.bytes)
            ? serializeSize(outputs[filename]?.bytes as number)
            : 'unknown'
        ]),
        ['', '\n']
      ] as string[][]

      printTable(message)
      return ExitCode.OK
    })
  )

  return handleWriteOutFiles$
}

export default build
