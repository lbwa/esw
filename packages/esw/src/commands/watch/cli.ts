import { ExitCode, stdout, isDef } from '@eswjs/common'
import { defer, NEVER, map, switchMap } from 'rxjs'
import arg from 'arg'
import omit from 'lodash/omit'
import { BuildOptions } from 'esbuild'
import { CommandRunner } from '../../cli/dispatch'
import { resolveArgv } from '../../cli/argv'
import { WatchArgsSpec, WATCH_ARGS_SPEC } from './cli-spec'
import runWatch from './node'

const watch: CommandRunner<ExitCode> = function (argv = []) {
  const handlePrintUsage$ = resolveArgv(
    argv,
    WATCH_ARGS_SPEC,
    defer(() => {
      stdout.raw(`
Description:
  watch files and run command when files change

Usage
  $ esw watch [options]
    
`)
      return NEVER
    })
  )

  const normalizedBuildArgs$ = handlePrintUsage$.pipe(
    map((args: arg.Result<WatchArgsSpec>) => {
      args['--absWorkingDir'] ??= process.cwd()
      const entryPoints = args._.filter(pending => !pending.startsWith('-'))
      args['--entryPoints'] ??= entryPoints.length > 0 ? entryPoints : undefined
      return omit(args, '_')
    }),
    map(args =>
      Object.keys(args).reduce((options, key) => {
        const value = args[key as keyof WatchArgsSpec]
        if (isDef(value)) {
          const name = key.replace(/^-+/, '') as keyof BuildOptions
          // @ts-expect-error mixed types
          options[name] = value as BuildOptions[keyof BuildOptions]
        }
        return options
      }, {} as BuildOptions)
    )
  )

  return normalizedBuildArgs$.pipe(
    switchMap(options => runWatch(options)),
    map(() => ExitCode.OK)
  )
}

export default watch
