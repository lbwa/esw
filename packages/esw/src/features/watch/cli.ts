import { ExitCode, stdout, isDef } from '@eswjs/common'
import { defer, NEVER, map, switchMap } from 'rxjs'
import arg from 'arg'
import omit from 'lodash/omit'
import { BuildOptions } from 'esbuild'
import { WatchArgsSpec, WATCH_ARGS_SPEC } from './cli-spec'
import runWatch from './node'
import { CommandRunner } from '@cli/dispatch'
import { resolveArgv } from '@cli/argv'

const watch: CommandRunner<ExitCode> = function (argv = []) {
  const handlePrintUsage$ = resolveArgv(
    argv,
    WATCH_ARGS_SPEC,
    defer(() => {
      stdout.raw(`
Description:
  watch working directory, run build again when files change

Usage
  esw watch [entry] [options]

  [entry] represents the library entry point.
          esw would run options inference when an entry files isn't explicitly specified.
          On the other hand, you should always specify a entry point explicitly when the main and module have a different basename.

  [options] esbuild options, see https://esbuild.github.io/

`)
      return NEVER
    })
  )

  const normalizedBuildArgs$ = handlePrintUsage$.pipe(
    map((args: arg.Result<WatchArgsSpec>) => {
      args['--absWorkingDir'] ??= process.cwd()
      const entryPoints = args._.filter(pending => !pending.startsWith('-'))
      args['--entryPoints'] ??= entryPoints.length > 0 ? entryPoints : undefined
      return omit(args, '_')
    }),
    map(args =>
      Object.keys(args).reduce((options, key) => {
        const value = args[key as keyof WatchArgsSpec]
        if (isDef(value)) {
          const name = key.replace(/^-+/, '') as keyof BuildOptions
          // @ts-expect-error mixed types
          options[name] = value as BuildOptions[keyof BuildOptions]
        }
        return options
      }, {} as BuildOptions)
    )
  )

  return normalizedBuildArgs$.pipe(
    switchMap(options => runWatch(options)),
    map(() => ExitCode.OK)
  )
}

export default watch
