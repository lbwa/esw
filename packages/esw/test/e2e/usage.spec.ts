import path from 'path'
import { format } from 'util'
import { createCliDriver } from './driver/cli'
import { createSandbox, Sandbox } from './sandbox/sandbox'

describe('cli stdout', () => {
  let sandbox: Sandbox
  beforeAll(async () => {
    sandbox = await createSandbox()
    await sandbox.load(path.resolve(__dirname, 'fixtures/no-options'))
    await sandbox.install({
      esw: format('file:%s', process.env.E2E_LIB_PACK_PATH)
    })
  })

  beforeEach(async () => {
    return sandbox.reset()
  })

  afterAll(async () => {
    return sandbox.terminate()
  })

  it.each([
    ['print global message', ['--help']],
    ['print global message by alias', ['-h']],
    ['print global message with unknown command', ['unknownCommand']],
    ['print build usage', ['build', '--help']],
    ['print build usage by alias', ['build', '-h']]
  ])('should %s', async (name, args) => {
    const driver = createCliDriver(sandbox.spawn('esw', args))
    const stdout = await driver.waitForStdout()
    expect(stdout).toMatchSnapshot(name)
  })

  it.each([
    ['print version', ['--version']],
    ['print version by alias', ['-v']]
    // eslint-disable-next-line @typescript-eslint/naming-convention
  ])('should %s', async (_, args) => {
    const driver = createCliDriver(sandbox.spawn('esw', args))
    const stdout = await driver.waitForStdout()
    expect(stdout).toMatch(/^v\d+\.\d+\.\d+/i)
  })
})
