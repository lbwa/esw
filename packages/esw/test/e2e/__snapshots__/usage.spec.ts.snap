// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli stdout should print build usage by alias: print build usage by alias 1`] = `
"
Description
  Compiles the codebase with options inference.

Usage
  esw build [entry] [options]

  [entry] represents the library entry point.
          esw would run options inference when an entry point isn't explicitly specified.
          On the other hand, you should always specify an entry point explicitly when the main and module have a different basename.

          Note that a part of inferred options didn't work once we found multiple entry points.

  [options] esbuild options, see https://esbuild.github.io/

"
`;

exports[`cli stdout should print build usage: print build usage 1`] = `
"
Description
  Compiles the codebase with options inference.

Usage
  esw build [entry] [options]

  [entry] represents the library entry point.
          esw would run options inference when an entry point isn't explicitly specified.
          On the other hand, you should always specify an entry point explicitly when the main and module have a different basename.

          Note that a part of inferred options didn't work once we found multiple entry points.

  [options] esbuild options, see https://esbuild.github.io/

"
`;

exports[`cli stdout should print global message by alias: print global message by alias 1`] = `
"
Usage
  $ esw <command>

Available commands
  build, watch

Options
  --version, -v      Show version number
  --help, -h         Display help messages

For more information run a command with the --help flag
  esw <available command> --help

"
`;

exports[`cli stdout should print global message with unknown command: print global message with unknown command 1`] = `
"
Usage
  $ esw <command>

Available commands
  build, watch

Options
  --version, -v      Show version number
  --help, -h         Display help messages

For more information run a command with the --help flag
  esw <available command> --help

"
`;

exports[`cli stdout should print global message: print global message 1`] = `
"
Usage
  $ esw <command>

Available commands
  build, watch

Options
  --version, -v      Show version number
  --help, -h         Display help messages

For more information run a command with the --help flag
  esw <available command> --help

"
`;
